### PR TITLE
Implement a provider model for RDS/Stdout mysql providers.

### PR DESCRIPTION
**Pull request by fubarhouse**
Based off the work done in #28, this segments the execution flow in a more coherent way using the 'provider model' to split out common functionality between the two routes of execution.

It keeps a single input for the user, where the user can provide a route of path, or use the default which is the current implementation on the main branch.

image

### CODE REVIEW

File: internal/mysql/providers/provider.go [hunk 1]
```diff
@@ -0,0 +1,20 @@
+package providers
+
+// MTKProvider is an interface implementing the required functionality for a Provider.
+type MTKProvider interface {
```
Comment (nickschuch) [inline internal/mysql/providers/provider.go:4]: Could be called "Interface". I think it's a bit better than MTKProvider.

File: internal/mysql/providers/provider.go [hunk 2]
```diff
@@ -0,0 +1,20 @@
+package providers
```
Comment (nickschuch) [inline internal/mysql/providers/provider.go:1]: Package should be `provider`

File: internal/mysql/providers/provider.go [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+
+// DumpParams is used to pass parameters to the Dump function.
+type DumpParams struct {
+	SelectMap          map[string]map[string]string
+	WhereMap           map[string]string
+	FilterMap          map[string]string
+	UseTableLock       bool
+	ExtendedInsertRows int
+
+	Provider string // Provider configuration.
```
Comment (nickschuch) [inline internal/mysql/providers/provider.go:17]: Do we need this?

File: internal/mysql/providers/provider.go [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+	DataPath string // S3 URI configuration for RDS provider.
```
Comment (nickschuch) [inline internal/mysql/providers/provider.go:18]: This can be set on the rds provider client

File: internal/mysql/providers/provider.go [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+	Region   string // Region configuration for RDS provider.
```
Comment (nickschuch) [inline internal/mysql/providers/provider.go:19]: This can be set on the rds provider client

File: internal/mysql/providers/stdout/provider.go [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+// NewClient for dumping a full or single table from a database.
+func NewClient(db *sql.DB, logger *log.Logger) *Client {
+	return &Client{
+		DB:     db,
+		Logger: logger,
+	}
+}
+
+// QueryColumnsForTable for a given table.
+func (d *Client) QueryColumnsForTable(table string, params providers.DumpParams) ([]string, error) {
```
Comment (nickschuch) [inline internal/mysql/providers/stdout/provider.go:28]: RDS and Stdout implement this with the exact same code.

File: internal/mysql/providers/rds/provider.go [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+// NewClient for dumping a full or single table from a database.
+func NewClient(db *sql.DB, logger *log.Logger) *Client {
+	return &Client{
+		DB:     db,
+		Logger: logger,
+	}
+}
+
+// QueryColumnsForTable for a given table.
+func (d *Client) QueryColumnsForTable(table string, params providers.DumpParams) ([]string, error) {
```
Comment (nickschuch) [inline internal/mysql/providers/rds/provider.go:28]: RDS and Stdout implement this with the exact same code.

**Comment by nickschuch** [comment]
Love the approach. I have left a review for next steps. Can we also get a functional test for atleast the `stdout` parameter? That way we know we aren't breaking our currently `.sql` output file implementation.

**Comment by fubarhouse** [comment]
Here's the functional test - all green.
https://github.com/skpr/mtk/actions/runs/9691644462/job/26743508199

**Pull request merged**